### PR TITLE
feat(treesitter): add injection language fallback

### PR DESCRIPTION
**Pull request by clason**
Remove the `#inject-language!` directive in favor of including the fallback to filetype in the injection logic itself. Also try the language name as lower case as a fallback, removing the need for downstream `#downcase!` directive. 

@lewis6991

### CODE REVIEW

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 1]
```diff
@@ -635,6 +635,14 @@ local function add_injection(t, tree_index, pattern, lang, combined, ranges)
   table.insert(t[tree_index][lang][pattern].regions, ranges)
 end
 
+-- TODO(clason): replace by refactored `ts.has_parser` API (without registering)
+---@param lang string
+---@return boolean # true if parser for {lang} exists on rtp
+local has_parser = function(lang)
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/languagetree.lua:641]:
```suggestion
local function has_parser(lang)
```

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 2]
```diff
@@ -652,7 +660,18 @@ function LanguageTree:_get_injection(match, metadata)
 
     -- Lang should override any other language tag
     if name == 'injection.language' then
-      lang = vim.treesitter.get_node_text(node, self._source, { metadata = metadata[id] })
+      local text =
+        vim.treesitter.get_node_text(node, self._source, { metadata = metadata[id] }):lower()
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/languagetree.lua:664]:
```suggestion
        vim.treesitter.get_node_text(node, self._source, { metadata = metadata[id] })
```

As mentioned, I'd much prefer that we didn't do this.

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 3]
```diff
@@ -652,7 +660,18 @@ function LanguageTree:_get_injection(match, metadata)
...
+      if not text then
+        break
+      elseif has_parser(text) then
+        lang = text
+      else
+        local alias = vim.treesitter.language.get_lang(text)
+        if alias and has_parser(alias) then
+          lang = alias
+        end
+      end
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/languagetree.lua:674]: Can this be factored into a function like `resolve_lang()`?

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 4]
```diff
@@ -652,7 +660,18 @@ function LanguageTree:_get_injection(match, metadata)
 
     -- Lang should override any other language tag
     if name == 'injection.language' then
-      lang = vim.treesitter.get_node_text(node, self._source, { metadata = metadata[id] })
+      local text =
+        vim.treesitter.get_node_text(node, self._source, { metadata = metadata[id] }):lower()
+      if not text then
+        break
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/languagetree.lua:666]: I don't think this possible. `get_node_text()` returns `string` in all paths.

The case it doesn't is when `metadata.text` is defined but isn't a string.

File: runtime/lua/vim/treesitter/languagetree.lua (hunk 2 above)
Comment (clason) [reply runtime/lua/vim/treesitter/languagetree.lua:664]: I still haven't heard a compelling argument that outweighs the significant benefits (no need for custom directives that make it harder to share queries).

Comment (lewis6991) [reply runtime/lua/vim/treesitter/languagetree.lua:664]: The benefits aren't significant IMO, and it implies that treesitter languages must be case-insensitive which isn't something we can decide or enforce.

Unless upstream makes this rule, we shouldn't do this.

Comment (clason) [reply runtime/lua/vim/treesitter/languagetree.lua]: Well, then it stands opinion against opinion.

Comment (lewis6991) [reply runtime/lua/vim/treesitter/languagetree.lua]: As a compromise we could try `lower()` as a fallback if nothing else resolves. That way we don't force case insensitivity.

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 5]
```diff
@@ -635,6 +635,29 @@ local function add_injection(t, tree_index, pattern, lang, combined, ranges)
...
+local resolve_lang = function(alias)
+  if has_parser(alias) then
+    return alias
+  else
+    local lang = vim.treesitter.language.get_lang(alias)
+    if lang and has_parser(lang) then
+      return lang
+    end
+  end
+end
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/languagetree.lua:659]: Leverage the `return`s to reduce nesting.

```suggestion
local function resolve_lang(alias)
  if has_parser(alias) then
    return alias
  end
 
  local lang = vim.treesitter.language.get_lang(alias)
  if lang and has_parser(lang) then
    return lang
  end
end
```

File: runtime/doc/news.txt [hunk 6]
```diff
@@ -120,8 +114,16 @@ The following new APIs and features were added.
...
-• Bundled treesitter parser and queries (highlight, folds) for Markdown,
-  Python, and Bash.
+• Treesitter
+  • Bundled parsers and queries (highlight, folds) for Markdown, Python, and
+    Bash.
+  • Added |vim.treesitter.query.omnifunc()| for treesitter query files (set by
+    default).
+  • |Query:iter_matches()| now has the ability to set the maximum start depth
+    for matches.
+  • `@injection.language` is now case insensitive and allows specifying a
```
Comment (lewis6991) [inline runtime/doc/news.txt:124]: This isn't quite true now.

```suggestion
  • `@injection.language` now has smarter resolution and will now fallback to language aliases and/or attempt lower case variants of the text.
```

**Review by lewis6991 — APPROVED** [review]

**Pull request merged**